src/main/scala/io/iohk/ethereum/forkid/ForkIdValidator.scala

```diff
@@ -0,0 +1,140 @@
+    val forks = ForkId.gatherForks(config)
+    validatePeer[F](genesisHash, forks)(currentHeight, remoteForkId)
+  }
+
+  private[forkid] def validatePeer[F[_]: Monad: Logger](
+      genesisHash: ByteString,
+      forks: List[BigInt]
+  )(currentHeight: BigInt, remoteId: ForkId): F[ForkIdValidationResult] = {
+    val checksums: Vector[BigInt] = calculateCheckchecksums(genesisHash, forks)
+
+    // find the first unpassed fork and it's index
+    val (unpassedFork, i) =
+      forks.zipWithIndex.find { case (fork, _) => currentHeight < fork }.getOrElse((maxUInt64, forks.length))
+
+    // The checks are left biased -> whenever a result is found we need to short circuit
+    val validate = (for {
+      _ <- liftF(Logger[F].trace(s"Before checkMatchingHashes"))
+      matching <- fromEither[F](
+        checkMatchingHashes(checksums, remoteId, currentHeight, i).toLeft("hashes didn't match")
+      )
+      _ <- liftF(Logger[F].trace(s"checkMatchingHashes result: $matching"))
+      _ <- liftF(Logger[F].trace(s"Before checkSubset"))
+      sub <- fromEither[F](checkSubset(checksums, forks, remoteId, i).toLeft("not in subset"))
+      _ <- liftF(Logger[F].trace(s"checkSubset result: $sub"))
+      _ <- liftF(Logger[F].trace(s"Before checkSuperset"))
+      sup <- fromEither[F](checkSuperset(checksums, remoteId, i).toLeft("not in superset"))
+      _ <- liftF(Logger[F].trace(s"checkSuperset result: $sup"))
+      _ <- liftF(Logger[F].trace(s"No check succeeded"))
+      _ <- fromEither[F](Either.left[ForkIdValidationResult, Unit](ErrLocalIncompatibleOrStale))
+    } yield ()).value
+
+    for {
+      _ <- Logger[F].debug(s"Validating $remoteId")
+      _ <- Logger[F].trace(s" list: $forks")
+      _ <- Logger[F].trace(s"Unpassed fork $unpassedFork was found at index $i")
+      res <- validate.map(_.swap)
+      _ <- Logger[F].debug(s"Validation result is: $res")
+    } yield (res.getOrElse(Connect))
+  }
+
+  private def calculateCheckchecksums(
+      genesisHash: ByteString,
+      forks: List[BigInt]
+  ): Vector[BigInt] = {
+    val crc = new CRC32()
+    crc.update(genesisHash.asByteBuffer)
+    val genesisChecksum = BigInt(crc.getValue())
+
+    genesisChecksum +: (forks.map { fork =>
+      crc.update(bigIntToBytes(fork, 8))
+      BigInt(crc.getValue())
+    }).toVector
+  }
+
+  /**
+    * 1) If local and remote FORK_HASH matches, compare local head to FORK_NEXT.
+    * The two nodes are in the same fork state currently.
+    * They might know of differing future forks, but that’s not relevant until the fork triggers (might be postponed, nodes might be updated to match).
+    * 1a) A remotely announced but remotely not passed block is already passed locally, disconnect, since the chains are incompatible.
+    * 1b) No remotely announced fork; or not yet passed locally, connect.
+    */
+  private def checkMatchingHashes(
+      checksums: Vector[BigInt],
+      remoteId: ForkId,
+      currentHeight: BigInt,
+      i: Int
```

Comment:
we could directly pass the current fork block hash instead of `i` and a list. It's not really obvious that the index of the next fork is actually also the index of the current checksums.